File: tui/internal/ui/baskets_test.go
package ui

import (
	"strings"
	"testing"

	"hejje.money/tui/internal/api"
)

func TestRenderBasketShowsLegsInPlacementOrderAndSplitsProgress(t *testing.T) {
	zero, one := 0, 1
	b := api.Basket{ID: "b1", Status: "ROLLED_BACK", Policy: "ALL_OR_NOTHING", Rollback: "CLOSE_FILLED_LEGS", Detail: "a leg failed; closed 1 of 1 filled leg(s)",
		Legs: []api.BasketLeg{
			{Sequence: 1, Side: "BUY", Quantity: 10, OrderType: "MARKET", Status: "ROLLED_BACK", ExecutionOrder: &one},
			{Sequence: 2, Side: "SELL", Quantity: 10, OrderType: "MARKET", HedgeFirst: true, Status: "FILLED", ExecutionOrder: &zero},
			{Sequence: 3, Side: "BUY", Quantity: 5, OrderType: "LIMIT", Status: "SKIPPED", Detail: "not placed: an earlier leg failed"}}}
	out := RenderBasket(b)
	for _, want := range []string{"BASKET b1  ROLLED_BACK  ALL_OR_NOTHING (rollback CLOSE_FILLED_LEGS)", "closed 1 of 1", "first", "not placed"} {
		if !strings.Contains(out, want) {
			t.Fatalf("missing %q in\n%s", want, out)
		}
	}
	if strings.Index(out, "SELL") > strings.Index(out, "ROLLED_BACK   ") && strings.Index(out, "2   SELL") > strings.Index(out, "1   BUY") {
		t.Fatalf("hedge leg should come first\n%s", out)
	}
	if !strings.Contains(RenderBaskets([]api.Basket{b}), "1/3") {
		t.Fatal("list render")
	}
	s := []api.SplitOrder{{ID: "s1", Side: "BUY", Quantity: 300, Filled: 100, Children: 1, Status: "EXPIRED", Detail: "deadline passed"}}
	if !strings.Contains(RenderSplits(s), "100/300 (1)") || RenderSplits(nil) != "No split orders.\n" {
		t.Fatal("splits render")
	}
}

File: web/tests/baskets.test.ts
import { describe, expect, it } from 'vitest';
import { legSummary, placementOrder, splitProgress } from '../src/lib/baskets';
import { Basket, BasketLeg, SplitOrder } from '../src/api/types';

const leg = (sequence: number, hedgeFirst: boolean, status: BasketLeg['status'], executionOrder?: number): BasketLeg => ({
  id: `l${sequence}`, sequence, hedgeFirst, executionOrder, instrumentId: 'i', side: 'BUY', quantity: 1, orderType: 'MARKET', product: 'MIS', status,
});

describe('basket and split helpers', () => {
  it('orders legs as placed: hedges first before placement, then the recorded order', () => {
    expect(placementOrder([leg(1, false, 'PENDING'), leg(2, true, 'PENDING'), leg(3, false, 'PENDING')]).map((l) => l.sequence)).toEqual([2, 1, 3]);
    expect(placementOrder([leg(1, false, 'FILLED', 1), leg(2, true, 'FILLED', 0)]).map((l) => l.sequence)).toEqual([2, 1]);
  });
  it('summarises fills and rollbacks', () => {
    const b = { legs: [leg(1, false, 'ROLLED_BACK'), leg(2, false, 'FAILED'), leg(3, false, 'SKIPPED')] } as Basket;
    expect(legSummary(b)).toBe('0/3 filled, 1 rolled back');
  });
  it('reports split progress', () => {
    const s = { quantity: 300, filledQuantity: 100, children: 1, policy: { maxChildQuantity: 100 } } as SplitOrder;
    expect(splitProgress(s)).toEqual({ pct: 33, text: '100/300 in 1 child order(s) of ≤ 100' });
  });
});
